### PR TITLE
option to read tiff metadata

### PR DESCRIPTION
**Pull request by jenniferdai1**
addressing issue #1940
Automatic inversion of binary image stored as TIFFILE

created imread_metadata, imread_with_metadata, and imread(include_metadata = False) methods in tifffile_plugin.py,  as suggested in a comment in the issue thread

added a new method, asarray_with_metadata, in tifffile.py that returns the image array along with a list of metadata dictionaries for each page. 
Returns metadata dictionary for first page by default, like how tifffile returns the first image series by default.

### CODE REVIEW

**Comment by stefanv** [comment]
Let's do:

`imread` : image only
`imread_metadata` : image + metadata

File: skimage/io/_plugins/tifffile_plugin.py [hunk 1]
```diff
@@ -13,7 +13,7 @@ def imread(fname, dtype=None, **kwargs):
     kwargs : keyword pairs, optional
         Additional keyword arguments to pass through (see ``tifffile``'s
         ``imread`` function).
-
+
```
Comment (jni) [inline skimage/io/_plugins/tifffile_plugin.py]: Clean up whitespace here and above

**Comment by jenniferdai1** [comment]
Sorry, could you explain what you mean by "separate" metadata reader? I was trying to follow your suggestion in the comment thread ("Some options:.....")

Thank you for your input!

**Comment by jni** [comment]
Hi Jennifer! First of all, thanks for tackling this, I'll be very happy to see it resolved!

Secondly, what I meant was, write a (potentially private) function that reads *only* metadata, and then say:

```python
def imread_metadata(filename):
    return imread(filename), _metadata(filename)
```

I think that would be cleaner than expanding the imread function itself.

Thirdly, although I agree with @stefanv that we should limit the number of options (from the Zen of Python: "There should be one-- and preferably only one --obvious way to do it."), I don't agree with the best choice here, and my preference would be `imread(filename, (return_|with_)metadata=False)`. This is not a strong opinion but rather a gut preference. The downside is the changing function signature depending on input, but it's a common enough pattern in SciPy that I don't think it's a big problem. Let's wait for a counterpoint from @stefanv about this before you keep cranking on this...

And finally, thanks for including a good test from the get-go, but I would like to specifically make sure that the case that started this (inverted values) works, so could you add:
- a test image for that use case, and
- a test of multipage metadata.

Thank you!

**Comment by stefanv** [comment]
@jni If we add the "metadata=False/True" flag, then I think we should not have imread_metadata implemented.  For me, it's one or the other.  I'm not a big fan of functions that use a flag to change the type of output they provide, hence my preference for ``imread_metadata``, but I wouldn't argue too much either way.

**Comment by jni** [comment]
@stefanv I agree that it's one or the other, and I also have only a small bias towards the kwarg. For me, it just seems more natural that this would be part of the imread function rather than a separate function, and I like to keep my tab-completion clean. =) Perhaps we can just have a vote? @emmanuelle @JDWarner @ahojnnes @blink1073 ?

For context, we are trying to decide whether to return image metadata using a separate function (`imread_metadata`, returning both the image and the metadata), or using a keyword argument to `imread`, so that it would optionally return metadata as well as the image, based on the value of the `metadata=True/False` kwarg.

Actually, I just thought of another argument in favour of the kwarg: the option to read metadata should probably generalise to `imread_collection`. That's easier with a kwarg.

**Comment by blink1073** [comment]
I prefer the approach taking by `imageio`:

"Reads an image from the specified file. Returns a numpy array, which comes with a dict of meta data at its ‘meta’ attribute."

http://imageio.readthedocs.org/en/latest/userapi.html#imageio.imread

**Comment by jni** [comment]
@blink1073 I like that idea! It actually crossed my mind but then I was like nooooo, are we even *allowed* to do that??? =P

**Comment by stefanv** [comment]
Remember the Image object that got deprecated around 0.9? :)

Let's stick to standard structures and go with the keyword argument.

**Comment by jni** [comment]
@stefanv very good point — super easy to lose the metadata when it's in some undocumented extra field.

**Comment by emmanuelle** [comment]
+1 for sticking with standard structures :-). I'd slightly prefer the separate function, but I don't have a strong dislike of the kw argument either.

File: skimage/io/_io.py [hunk 2]
```diff
@@ -73,6 +74,68 @@ def imread(fname, as_grey=False, plugin=None, flatten=None,
...
+    metadata : list
+        A list containing TIFF pages' metadata dictionaries.
+
+    Other parameters
+    ----------------
+    plugin_args : keywords
+        Passed to the given plugin.
+
+    """
+    # Backward compatibility
```
Comment (AbdealiJK) [inline skimage/io/_io.py]: Is backward compatibility needed for a completely new function too ?

**Comment by AbdealiJK** [comment]
So, I notice that the `tag` attribute from `tifffile` is directly given out.
Wouldn't it be better to curate the list of metadata that's being shown ? This way it can be standardized for other plugins too. Otherwise, the user will have to read the doumentation of each backend individually and also decipher which backend was used in the file to be able to use the metadata (as otherwise the keys will change).

The metadata that is exposed by the function can be added based on requirement, otherwise `skimage` which s an image processing toolkit would be bogged down by exif/meta data reading.

Also, the tiff tags mentioned at [tiff tags](http://www.digitalpreservation.gov/formats/content/tiff_tags.shtml) have: Comments, Model (The model of the scanner), Artist, TargetPrinter, XMP, BatteryLevel, Copyright, ICCProfile, GPSInfo, and a whole lot of information that an image processing toolkit wouldn't care about. This seems like unnecessary noise. To read that sort of meta-information, the user should actually use an exif/meta reader like exiftool, gexiv2, etc.

**Comment by stefanv** [comment]
@AbdealiJK Agreed that we should have a set of standard meta-data, and then perhaps a custom or raw entry that contains everything the reader can provide.

**Comment by stefanv** [comment]
See https://github.com/scikit-image/scikit-image/issues/2709

**Pull request closed without merging**